eth_abi/registry.py

```diff
@@ -0,0 +1,146 @@
+from collections import namedtuple
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class ExistingEntryFound(ABIRegistryError):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+PredicateItem = namedtuple('PredicateItem', ('predicate', 'value'))
+
+
+def ensure_changed(ExcClass, error_msg):
+    def decorator(old_method):
+        def new_method(self, *args, **kwargs):
+            len_before = len(self._items)
+            result = old_method(self, *args, **kwargs)
+            len_after = len(self._items)
+
+            if len_before == len_after:
+                raise ExcClass(error_msg)
+
+            return result
+
+        return new_method
+
+    return decorator
+
+
+class PredicateList:
+    def __init__(self):
+        self._items = []
+
+    def add(self, predicate, value):
+        self._items.insert(0, PredicateItem(predicate, value))
+
+    def match(self, *args, **kwargs):
+        return tuple(i.value for i in self._items if i.predicate(*args, **kwargs))
+
+    @ensure_changed(KeyError, 'Given predicate not found')
+    def remove_exact(self, predicate):
+        self._items = [i for i in self._items if i.predicate is not predicate]
+
+    @ensure_changed(KeyError, 'No matching items found')
+    def remove_matching(self, *args, **kwargs):
+        self._items = [i for i in self._items if not i.predicate(*args, **kwargs)]
+
+    def __contains__(self, item):
+        return item in tuple(i.predicate for i in self._items)
+
+
+class ABIRegistry:
+    def __init__(self):
+        self._encoder_registry = {}
+        self._decoder_registry = {}
+
+        self._encoder_predicates = PredicateList()
+        self._decoder_predicates = PredicateList()
+
+    def register_encoder(self, lookup, encoder):
+        """
+        Registers an encoder for the given type lookup.
+        """
+        if callable(lookup):
+            # Is predicate lookup
+            self._encoder_predicates.add(lookup, encoder)
+            return
+
+        # Is string lookup
+        if lookup in self._encoder_registry:
```

Comment:
We've generally tried to write our conditionals to be explicit in the following manner.

```python
if callable(thing):
    # do something
elif isinstance(thing, str):
    # do something
else:
    raise TypeError("<good error message indicating type not supported, what type was given, what types are allowed")
```

It ensures we don't hit weird cases where someone uses `bytes` or an `int` or something and our registry ends up with weird values.

Reply:
Sounds good.  Just FYI, part of my deciding to do things that way had to do with similar duck-typing I saw here:
https://github.com/ethereum/eth-abi/blob/master/eth_abi/abi.py#L26-L29

Should we consider being more explicit in that case as well?

Reply:
Yes, that would be an improvement.  Adding type hints here (and mypy typing would also help but that should probably be done independently).